Support tuples with ghost components

fpoli opened the pull request:
This PR adds support for tuples with ghost components. In addition to https://github.com/secure-foundations/dafny/commit/c51154653ec55b2a08cd3610228f00968250e989, authored by @Chris-Hawblitzel, the major changes are:
- The `ghost` keyword is not allowed in the patterns of match statements.
- The parser handles ambiguous cases like `u := F( a < (b, ghost b), c > (d) );`
- `ghost` can be added to tuples with named arguments. For example `(1 := 123, ghost 0 := 234)`, which is equivalent to `(ghost 234, 123)`.
- The `ghost` keyword is pretty-printed.
- Various fixes to the compilers, such that ghost components of tuples don't show up in the target language. In particular, unlike with user-defined datatypes, ghost components of tuples don't cause the generation of type parameter in C#.

keyboardDrummer commented on Source/Dafny/Dafny.atg:
```diff
@@ -4105,13 +4133,22 @@ CasePatternLocal<.out CasePattern<LocalVariable> pat, bool isGhost.>
     ")"                                (. // Parse parenthesis without an identifier as a built in tuple type.
-                                          theBuiltIns.TupleType(id, arguments.Count, true); // make sure the tuple type exists
+                                          // TODO(fpoli): Why is the following call needed?
+                                          // TODO(fpoli): Given the statement`ghost var (x, ghost y) := ...;`, should we consider the type of the tuple to be `(ghost, ghost)` or `(non-ghost, ghost)`?
```
I would make this illegal since adding a redundant `ghost` implies the programmer didn't understand what they were doing.

fpoli replied:
It sounds good to me. I can easily generate a `SemErr(..)` for those cases.

RustanLeino replied:
I thought about this some more. Now, I don't think it's all that tricky. In fact, I don't think there's any difference from how things work today. Today, the `match` source expression is type checked. Then, the type inference machinery makes an effort to figure out the type of this expression (which it does by calling `PartiallySolveTypeConstraints()` at that time). If it figures out the type and it's some kind of tuple type, then it known what type to look up the constructors in. If it can't figure it out, it gives an error. So, I expect that we'll always know what type to look up the constructor in. (We may need to add a call to `PartiallySolveTypeConstraints()` in the resolution of `var (x, y) := e;` statements and `var (x, y) := e0; e1` expressions. I don't know if those calls are there today.)

fpoli replied:
`ghost var (x, ghost y) := ...;` is now a syntax error, as the inner ghost is not allowed by the grammar.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
```
Consider using syntax like `ghostArguments = ghostArguments ?? ..`

fpoli replied:
No, thanks for pointing that out. I'm new to C#.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
+        ghostArguments = new bool[dims].ToList().ConvertAll(_ => false);
```
Consider using `new bool[dims].Select(_ => false).ToList();` instead.

My impression is that `ConvertAll` has been replaced by `Select` in practice, so most C# devs won't know `ConvertAll` and it's better to avoid it. [More information](https://stackoverflow.com/a/1571843/93197)

RustanLeino replied:
`ConvertAll` has the advantage that it returns a `List`, so there's no need to follow it with `.ToList()` (and the implementation then immediately knows how much storage to allocate for the `List` to be created).

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       if (!tupleTypeDecls.TryGetValue(dims, out tt)) {
         Contract.Assume(allowCreationOfNewType);  // the parser should ensure that all needed tuple types exist by the time of resolution
-        if (dims == 2) {
+        if (dims == 2 && !ghostArguments.Contains(true)) {
```
I'm not sure, but shouldn't this be `if (dims - ghostArguments.Count(x => x) == 2)` ? I'm guessing the runtime instance of a tuple is the sum of all the non-ghost arguments in the compile-time tuple.

keyboardDrummer replied:
Shall we refactor this to
```
var attributes = dims == 2 ? DontCompile() : null;
tt = new TupleTypeDecl(ghostArguments, SystemModule, attributes);
```
?

fpoli replied:
> I'm not sure, but shouldn't this be `if (dims - ghostArguments.Count(x => x) == 2)` ? I'm guessing the runtime size of a tuple is the sum of the non-ghost arguments in the compile-time tuple.

I still have to look at how to compile the ghost components. The branch from which I'm taking most of the work defines different runtime names for tuples that have the same number of concrete components but different number of ghost components, and I'm not sure why.

RustanLeino replied:
At run time, each n-tuple with m ghost components will use an (n-m)-tuple (except when n-m is 1, in which case I suggest the compiler omit the tuple wrapper altogether). However, I think the enclosing `TupleType` method here needs to generate tuple definitions that _with_ the indicated ghost components.

fpoli replied:
As Runstan wrote, I think we need to declare different tuples for `(Y, Z)` and `(ghost X, Y, Z)`.

keyboardDrummer commented on Source/Dafny/Dafny.atg:
```diff
@@ -4105,13 +4133,22 @@ CasePatternLocal<.out CasePattern<LocalVariable> pat, bool isGhost.>
       ]
     ")"                                (. // Parse parenthesis without an identifier as a built in tuple type.
-                                          theBuiltIns.TupleType(id, arguments.Count, true); // make sure the tuple type exists
+                                          // TODO(fpoli): Why is the following call needed?
```
I'm not sure I understand the question, but `theBuiltIns.TupleType` records which dimensions of tuple types are created in the input program, which causes extra tuple type definitions to be emitted if needed.

fpoli replied:
To expand the question, I'm wondering why the parser needs to emit tuple definitions. I expected the parser to be a component that takes a program and just emits an AST and/or parsing errors, but this doesn't seem the case.

I'll try to comment that out and see what happens, thanks.

fpoli replied:
All tests still pass. I also found somewhere else a comment stating that "the parser should ensure that all needed tuple types exist by the time of resolution", which answers my question.

fpoli replied:
> Remy's observation has some merit. However, I wonder what would happen in an _incorrect_ program that tried something like:
> 
> ```
> var (x, y) := 100;
> ```
> 
> This is not type correct, but it _may_ be that the type checker expects the 2-tuple to exist at this time.

In both cases (before this PR and after commenting the `theBuiltIns.TupleType(..)`) dafny reports the following error. So, it's unclear whether the `theBuiltIns.TupleType(..)` is actually needed.

```
Error: type of RHS (int) does not match type of bound variable '_#Make2'
```

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
+      if (System.Char.IsDigit(s[s.Length - 1])) {
+        return ghostArguments.ConvertAll(_ => false);
+      } else {
+        return ghostArguments.Select((_, i) => s[s.Length - count + i]).ToList().ConvertAll(IsGhostFromChar);
```
This expression can be simplified

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
+        ? "" : String.Concat(ghostArguments.ConvertAll(IsGhostToChar)));
+    }
+
+    public static List<bool> GhostArgumentsFromString(string s, int count) {
```
Consider returning an `IEnumerable<string>`GhostArgumentsFromString, that way the caller can decide whether to do the extra work of allocating a new List.

If you look at the call site of this method, you can see the return value is immediately passed to `Zip` so storing the bool in a List was not needed.

```
        List<bool> ghostArguments = BuiltIns.GhostArgumentsFromString(Name, TypeArgs.Count);
        return "(" + Util.Comma(System.Linq.Enumerable.Zip(TypeArgs, ghostArguments), ...
```

fpoli replied:
I changed the types to `IEnumerable<bool>`.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
+      return c == 'G';
+    }
+
+    public static string GhostArgumentsToString(List<bool> ghostArguments) {
```
Is this string user-facing? We could make it more or less complex. Less complex would be to always return `String.Concat(ghostArguments.ConvertAll(IsGhostToChar)))` even if there are no ghosts, and more complex would be to return `3G2O` instead of `5GGGOO`.

This looks good to me though, just wanted to mention some alternatives.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,67 @@ public class BuiltIns
     }
+
+    public static string TupleTypeName(List<bool> ghostArguments) {
+      // TODO(fpoli): Why doesn't the TupleTypeName encode the type of the arguments, but (before this PR) just its length?
```
I'm guessing TupleTypeName is just the name of the runtime type of the Tuple, which is a generic type that can be used with multiple argument types.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4751,7 +4798,7 @@ private TupleTypeDecl(ModuleDefinition systemModule, List<TypeParameter> typeArg
     public override string CompileName {
       get {
-        return "Tuple" + Dims;
+        return BuiltIns.TupleTypeName(GhostArguments).Replace("_tuple#", "Tuple");
```
Do we need this change?

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4751,7 +4798,7 @@ private TupleTypeDecl(ModuleDefinition systemModule, List<TypeParameter> typeArg
     public override string CompileName {
       get {
-        return "Tuple" + Dims;
+        return BuiltIns.TupleTypeName(GhostArguments).Replace("_tuple#", "Tuple");
```
I think for the runtime name, only the non-ghost arguments matter, so the type would be `"Tuple" + GhostArguments.Count(x => !x)`

RustanLeino replied:
Correct. And furthermore, I think we should work on the compilers so that `_tuple#1` is never needed.

fpoli replied:
Changing the implementation to `"Tuple" + NonGhostDims` causes the errors reported below when running `dafny /compile:0` on the following program.

```
// RUN: %dafny /compile:0 /dprint:- "%s" /env:0 > "%t"
// RUN: %diff "%s.expect" "%t"

method Test()
{
    // Tuples with a single ghost component are allowed and they are compiled to empty tuples.
    var x := (ghost 123);
    var (y) := (ghost 123);
    ghost var z: (ghost int) := (ghost 123);
}
```

```
(0,-1): Error: more than one declaration of function name: Tclass._System.Tuple0
(0,-1): Error: more than one declaration of variable name: Tagclass._System.Tuple0
(0,-1): Error: more than one declaration of function name: $IsA#_System.Tuple0
(0,-1): Error: more than one declaration of function name: _System.Tuple0#Equal
(0,-1): Error: more than one declaration of variable name: class._System.Tuple0
5 name resolution errors detected in /var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl

*** Encountered internal translation error - re-running Boogie to get better debug information

/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2829,9): Error: more than one declaration of function name: Tclass._System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2831,0): Error: more than one declaration of variable name: Tagclass._System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2857,9): Error: more than one declaration of function name: $IsA#_System.Tuple0
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2870,9): Error: more than one declaration of function name: _System.Tuple0#Equal
/var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl(2882,0): Error: more than one declaration of variable name: class._System.Tuple0
5 name resolution errors detected in /var/folders/p3/fp88646n1_d92mg86l0knmz00000gs/T/AllowedGhost__module.bpl
```

It seems that each `TupleTypeDecl` generates its own Boogie variables  and functions based on `CompileName`.

RustanLeino replied:
I see. That's a bug. That is, `Translator.cs` should not use `.CompileName`. The Boogie names having to do with 2-tuples should look something like `#_System._tuple#2`, whereas the compile names may look like `_System.Tuple2`. It seems the verifier currently uses a mix of these names, which is not right.

I think it will be easiest for the verifier to use one different name for each type. That is, the tuple types `(,)` and `(ghost,)` would in Boogie have names like `_tuple#2oo` and `..._tuple#2go`. There will also be a Boogie name for the 1-tuple `(ghost)`.

The compiler can use names like `"Tuple" + NonGhostDims`. Eventually, we should optimize away `Tuple1`, but that can be a separate pull request.

fpoli replied:
I should have fixed the bug. There was a single call to `ctor.CompileName` in `Resolver.cs` that I changed to `ctor.Name`.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4694,29 +4734,36 @@ public enum ES { NotYetComputed, Never, ConsultTypeArguments }
-    public readonly int Dims;
+    public readonly List<bool> GhostArguments;
+
+    public int Dims
```
This field is used in `Compiler-java.cs` and I think there it means the runtime dimensions of the tuple type. Consider introduced a `NonGhostDims` field and seeing which usages of `Dims` should use that.

fpoli replied:
I added `NonGhostDims` with a TODO to check the usages of `Dims`.

keyboardDrummer commented on Test/ghost/PhantomData.dfy:
```diff
@@ -0,0 +1,63 @@
+    var (ghost y12, ghost y13) := b.1;
+}
+
+// TODO(fpoli): The following signature causes the error "Undeclared top-level type or type parameter: _tuple#2OG". What does it mean?
```
Seems like a runtime artefact. Maybe it goes away if we change the TupleTypeName above to `$"_tuple#{NonGhostDims}"`

fpoli replied:
I'll try, thanks. I didn't expect dafny with `/compile:0` to generate runtime stuff.

RustanLeino replied:
It seems to me that `theBuiltIns.TupleType` has not been called for `new bool[2] { false, true }`.

fpoli replied:
I fixed the bug.

keyboardDrummer commented on Test/ghost/PhantomData.dfy:
```diff
@@ -0,0 +1,63 @@
+
+method GhostBad3()
+{
+    var (ghost x) := 123; // TODO(fpoli): Is it fine to disallow this?
```
Is `var (x) := 123` even allowed? Would that be the same as `var x := 123` ?

fpoli replied:
It's probably disallowed, but I don't get a nice error message:

```
// Program.dfy

method Test()
{
  var (x) := 123;

  assert x == 123;
}
```

```
Program.dfy(7,8): Error: type of RHS (int) does not match type of bound variable '_#Make1'
```

RustanLeino replied:
`var (x) := 123` is currently not allowed, because there is no 1-tuple. However, if we support 1-tuples where the component is ghost (which I suggest we do), then we would support `var (x) := (ghost 123)`, or more generally `var (x) := e` where `e` has type `(ghost T)` for some `T`.

fpoli replied:
I'm going to consider `var (ghost x) := 123` a type error: the LHS is an empty tuple but the RHS an integer.

keyboardDrummer replied:
I'm confused by `var (x) := 123`, because the right-hand side doesn't even look like a singleton tuple, it just looks like a non-tuple value. I don't see how you would enable singleton tuples since `(123)` just looks `123` nested in a parenthesis expression.

fpoli replied:
I don't think we want to allow `var (x) := (123)`. What we want is `var (x) = (ghost 123)`, which currently parses fine.

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -4322,13 +4359,17 @@ ActualBindings<.List<ActualBinding> bindings.>
   ]
   Expression<out e, true, true>
-  (. binding = new ActualBinding(id, e); .)
+  (. binding = new ActualBinding(id, e, isGhost); .)
```
This is good. The `ghost` keyword should be allowed only if the `ActualBinding` denotes a component of a tuple. The Resolver should give an error if `ghost` is used for any other purpose. (Alternatively, you can add a parameter `bool allowGhostKeyword` to the `ActualBinding` grammar production and check the condition it here in the parser.)

fpoli replied:
I moved the `[ "ghost" ]` to `TupleArgs`. `ActualBinding` now has `isGhost` as a parameter.

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -2037,21 +2039,31 @@ TypeAndToken<out IToken tok, out Type ty, bool inExpressionContext>
                                   .)
-  [ Type<out ty>                  (. tupleArgTypes.Add(ty); .)
-    { "," Type<out ty>            (. tupleArgTypes.Add(ty); .)
+  [ [ "ghost"                     (. isGhost = true; .)
+    ]
+    Type<out ty>                  (. tupleArgTypes.Add(ty); ghostArguments.Add(isGhost); .)
+    { ","                         (. isGhost = false; .)
+      [ "ghost"                   (. isGhost = true; .)
+      ]
+      Type<out ty>                (. tupleArgTypes.Add(ty); ghostArguments.Add(isGhost); .)
```
There needs to be a corresponding change to the `_openparen` case of method `IsNonArrowType` in `Dafny.atg`. See the comment in method `IsGenericInstantiation` in `Dafny.atg`, and also see module `M1` in `Test/dafny0/UserSpecifiedTypeParameters.dfy`.

fpoli replied:
I see, the `IsTypeSequence` method needs to match `[ "ghost" ] Type { "," [ "ghost" ] Type }` instead of `Type { "," Type }`.

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -4322,13 +4359,17 @@ ActualBindings<.List<ActualBinding> bindings.>
-= (. IToken id = null; Expression e; .)
+= (. IToken id = null; Expression e; bool isGhost = false; .)
+  [
+    // TODO(fpoli): The linear-types branch uses `IF(la.kind == _ghost && !IsMatch() && !IsVar())` here. Any clue why they needed that?
```
> I'm confused. I thought the linear-types branch is from before `ActualBinding` was introduced. Has that branch been updated? (If so, I must be looking at the wrong repo.)

No, you looked at the right repo. That branch doesn't include `ActualBinding`. However, they define a `TupleArgs` rule that calls `Expression`:

```
TupleArgs<.List<Expression> args, List<Usage> usages.>
= (. Expression e; Usage usage = Usage.Ordinary; .)
  [ IF(la.kind == _linear && !IsMatch() && !IsVar()) "linear" (. usage = Usage.Linear; .) ]
  [ IF(la.kind == _ghost && !IsMatch() && !IsVar()) "ghost"   (. usage = Usage.Ghost; .) ]
  Expression<out e, true, true>                      (. args.Add(e); usages.Add(usage); .)
  { ","                                              (. usage = Usage.Ordinary; .)
    [ IF(la.kind == _linear && !IsMatch() && !IsVar()) "linear" (. usage = Usage.Linear; .) ]
    [ IF(la.kind == _ghost && !IsMatch() && !IsVar()) "ghost"   (. usage = Usage.Ghost; .) ]
    Expression<out e, true, true>                    (. args.Add(e); usages.Add(usage); .)
  }
  .
```
 which I think should become
```
TupleArgs<.List<ActualBinding> args.>
= (. ActualBinding binding; .)
  ActualBinding<out binding>    (. args.Add(binding); .)
  { ","
    ActualBinding<out binding>  (. args.Add(binding); .)
  }
  .
```
moving the `[ "ghost" ]` to the definition of `ActualBinding`.

> To your question, I understand `!IsVar()` -- it's to be able to distinguish between `(ghost var x := 5; x + x)` and `(ghost x)`, both of which begin with `(` `ghost`. We will need that in the same way. 

The example is very useful, thanks.

fpoli replied:
I moved the `[ IF ... ]` to the `TupleArgs`.

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -4061,13 +4080,19 @@ CasePattern<.out CasePattern<BoundVar> pat.>
                                        .)
-      [ CasePattern<out pat>           (. arguments.Add(pat); .)
-        { "," CasePattern<out pat>     (. arguments.Add(pat); .)
+      [ [ "ghost"                      (. isGhost = true; .)
+        ]
+        CasePattern<out pat>           (. arguments.Add(pat); ghostArguments.Add(isGhost); .)
+        { ","                          (. isGhost = false; .)
+          [ "ghost"                    (. isGhost = true; .)
```
Does it mean that it's fine to call `theBuiltIns.TupleType(id, arguments.Count, true, ghostArguments);` with the number of arguments instead of the number of ghost arguments, or should I try to remove it? I don't understand why the parser has to emit these tuple types. I expect the types to be fully known only after type checking rather than at parsing time.

fpoli replied:
Removing the `theBuiltIns.TupleType(id, arguments.Count, true, ...);` call doesn't seem to cause big problems in the test suite. Just a few tests fail because of unexpected C++ warnings, but they might be caused by other reasons. I'm debugging.

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -4105,13 +4133,22 @@ CasePatternLocal<.out CasePattern<LocalVariable> pat, bool isGhost.>
+        CasePatternLocal<out pat, isGhost || ghostArg>   (. arguments.Add(pat); ghostArguments.Add(isGhost || ghostArg); .)
+        { ","                                            (. ghostArg = false; .)
+          [ "ghost"                                      (. ghostArg = true; .)
+          ]
```
I removed the ghost keyword from `CasePatternLocal`. Now writing `var (x, ghost y) := ...` is a syntax error.

fpoli commented:
It seems that in Compiler-java.cs the generic code that compiles a DatatypeValue (supertype of tuples) needs to be special cased to skip type arguments for ghost components. That is, dafny already compiles datatypes with N non-ghost arguments and M ghost to classes with N+M type parameters, but in the case of tuples we want to compile to classes with just the N non-ghost type parameters.

fpoli commented:
Where is the code that emits the types (and type parameters) of method arguments in the Java compiler?

```dafny
// Dafny
method Test0(t0: (ghost int)) { ... }
```
Compilation:
```java
// Java
public class __default {
  public __default() {
  }
  public static void Test0(dafny.Tuple0<java.math.BigInteger> t0)
  //                                   ^^^^^^^^^^^^^^^^^^^^^^---- This is what I'm trying to fix. Where is it generated?
  {
    ...
  }
}
```

fpoli commented:
@RustanLeino Should the type parameters for ghost component show up in the target language of the dafny compilation?

Consider the `datatype PhantomData<T> = PhantomData(ghost value: T)` type. In Java and C# it is compiled to a class with a `<T>` type parameter for the ghost argument, even if the argument is unused in the program.

Now, consider the `(ghost T)` tuple type, which should be compiled to an empty tuple. We have some options:
1. If we consider the tuple to be a special case of a datatype (in fact, `TupleTypeDecl` extends `DatatypeDecl`) then `(ghost T)` should be compiled to a class `Tuple1G<T>` where the type `T` is unused. To do this, the generation of tuple class declarations in the Java compiler should be changed to generate ghost type parameters.
2. If we want to avoid generating unnecessary classes, `(ghost T)` should be compiled to a class `Tuple0` with no type parameters. To do this, the generic generation of datatype class declarations in the C# compiler should be fixed to special-case the tuple types.
 3. In addition to option 2, we can also change the compilation of datatypes like `datatype PhantomData<T> = PhantomData(ghost value: T)` to generate classes without type parameters for the ghost components. This would produce the best target code in my opinion, but it requires more implementation effort.

RustanLeino commented:
@fpoli Very good points.

It would not be correct (in general) to remove the type parameter of user-defined datatypes like `PhantomData` above. The reason is that such a type could have a member that makes use of the type parameter. So, we should not do option 3.

I think we should aim for option 2. That is, I think it is worthwhile to special-case tuple types to remove the type arguments that are not needed. Since tuples are built in, they are known not to have any compiled members that need the type parameters corresponding to the ghost arguments.

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -3820,17 +3837,45 @@ ParensExpression<out Expression e>
        }
        e = new ParensExpression(x, args[0].Actual);
      } else {
+       // Compute the actual position of ghost arguments
+       // TODO(fpoli): Is this the correct place to do this? Should this be moved to the resolver?
```
I have a question here.

fpoli replied:
Sorting is not needed. The argument name can be parsed and used as an index (if in range).

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -4486,6 +4529,24 @@ Expressions<.List<Expression> args.>
+TupleArgs<.List<ActualBinding> args.>
+= (. ActualBinding binding; bool isGhost = false; .)
+  // The IF is to distinguish between `(ghost var x := 5; x + x)` and (ghost x), both of which begin with `( ghost`.
+  // TODO(fpoli): It seems the linear-types branch also allows ghost match constructs, which is why they also need
+  // !IsMatch(). Should we do the same?
```
Actually, I'm not sure what a ghost match is. Do you know any example that might need a `ghost` in correspondence with a `match`? If not, I'll just remove the TODO.

fpoli replied:
We shouldn't need it. Remove.

fpoli commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4698,35 +4732,49 @@ public enum ES { NotYetComputed, Never, ConsultTypeArguments }
   public class TupleTypeDecl : IndDatatypeDecl
   {
-    public readonly int Dims;
+    // TODO(fpoli): `IsGhostArguments` should be read as "it's the isGhost information for the arguments", but it
+    // might sound grammatically incorrect. An alternative is `ArgumentsGhostness`.
+    public readonly List<bool> IsGhostArguments;
```
Is the `IsGhostArguments` name fine? I don't have a strong preference.

RustanLeino replied:
How about `ArgumentGhostness`?

fpoli replied:
`ArgumentGhostness`.

fpoli commented:
@RustanLeino I addressed the review and rebased on master.

fpoli commented:
The CI failure of the "Create release" step looks like a glitch:
```
   Segmentation fault (core dumped)
/home/runner/work/dafny/dafny/dafny/Source/Dafny/DafnyPipeline.csproj(4,5): error MSB3073: The command "dotnet tool restore" exited with code 139.
  Successfully created package '/home/runner/work/dafny/dafny/dafny/Binaries/DafnyRuntime.1.1.0.nupkg'.
failed! (Is Dafny or the Dafny server running?)
```

I'll force-push to restart the tests.

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -3821,17 +3839,51 @@ ParensExpression<out Expression e>
        }
        e = new ParensExpression(x, args[0].Actual);
      } else {
+       // Compute the actual position of ghost arguments
+       List<bool> argumentGhostness;
+       if (args.TrueForAll(arg => arg.FormalParameterName != null)) {
+         bool[] ghostness = new bool[args.Count];
+         bool[] initialized = new bool[args.Count];
+         for (int i = 0; i < initialized.Length; ++i) {
+           initialized[i] = false;
+         }
+         foreach (var arg in args) {
+           var success = int.TryParse(arg.FormalParameterName.val, out var index);
+           if (!success) {
+             SemErr(arg.FormalParameterName, "the name of tuple arguments must be an integer");
+           } else {
+             if (index < 0 || ghostness.Length <= index) {
+               SemErr(arg.FormalParameterName, "the name of tuple argument must describe a valid position in the tuple, starting from zero");
+             } else {
+               if (initialized[index]) {
+                 SemErr(arg.FormalParameterName, "different tuple arguments cannot have the same name");
+               } else {
+                 ghostness[index] = arg.IsGhost;
+                 initialized[index] = true;
```
This is too specific. The code here only needs to find the right tuple types _for correctly given tuples_. If parameters are duplicated or misspelled, the Resolver will complain later. I suggest the following logic:

```
var ghostness = new bool[args.Count];
for (var i = 0; i < args.Count; i++) {
  ghostness[i] = true;
}
for (var i = 0; i < args.Count; i++) {
  var arg = args[i];
  if (arg.IsGhost) {
    if (arg.FormalParameterName == null) {
      ghostness[i] = true;
    } else {
      var success = int.TryParse(arg.FormalParameterName.val, out var index);
      if (success && 0 <= index && index < args.Count) {
        ghostness[index] = true;
      }
    }
  }
}
var argumentGhostness = ghostness.ToList();
```

RustanLeino replied:
Note, btw, that a parameter with unnecessary leading zeros (e.g., `007`) is incorrect. `TryParse` will succeed, but that's okay, because the Resolver will find the error later.

fpoli replied:
Changed, thanks.

RustanLeino commented on Source/Dafny/Dafny.atg:
```diff
@@ -4489,6 +4539,22 @@ Expressions<.List<Expression> args.>
+/*------------------------------------------------------------------------*/
+TupleArgs<.List<ActualBinding> args.>
+= (. ActualBinding binding; bool isGhost = false; .)
+  // The IF is to distinguish between `(ghost var x := 5; x + x)` and (ghost x), both of which begin with `( ghost`.
```
Good catch!

RustanLeino commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -114,7 +114,7 @@ public class BuiltIns
     public readonly Dictionary<int, SubsetTypeDecl> PartialArrowTypeDecls = new Dictionary<int, SubsetTypeDecl>();  // same keys as arrowTypeDecl
     public readonly Dictionary<int, SubsetTypeDecl> TotalArrowTypeDecls = new Dictionary<int, SubsetTypeDecl>();  // same keys as arrowTypeDecl
-    readonly Dictionary<int, TupleTypeDecl> tupleTypeDecls = new Dictionary<int, TupleTypeDecl>();
+    readonly Dictionary<object, TupleTypeDecl> tupleTypeDecls = new Dictionary<object, TupleTypeDecl>();
```
Can `object` be more specific? For example, is it always a list of booleans?

Will `Dictionary` understand how to compare these keys?

If something else is needed as a key, an alternative would be to treat the array of booleans as a list of binary digits. Then prepend a `1` to those digits and treat as a `BigNumber`. For example, `('a', ghost 'b', ghost 'c')` would be represented by the binary number `1011` (that is, `11` in decimal).

Oh, or better yet, use the strings you already use, like `OGG` for the example above.

fpoli replied:
This code comes from the linear-types branch. They use nested tuples to index the dictionary by a sequence of booleans because C#'s `List` class doesn't provide the correct hash and equality methods.

If we wish to use `List<bool>` for the keys, we can pass to the constructor of the dictionary a comparer like the following (from [here](https://stackoverflow.com/a/10020580/2491528)).
```c#
class ListComparer<T> : IEqualityComparer<List<T>>
{
    public bool Equals(List<T> x, List<T> y)
    {
        return x.SequenceEqual(y);
    }

    public int GetHashCode(List<T> obj)
    {
        int hashcode = 0;
        foreach (T t in obj)
        {
            // etc
        }
        return hashcode;
    }
}
```

Using a `BigNumber` for the keys would also work. Constructing a key would be slower, but the hash and equality operations might be faster.

fpoli replied:
I changed the dictionary to use a `List<bool>` with a custom comparator.
```c#
readonly Dictionary<List<bool>, TupleTypeDecl> tupleTypeDecls = new Dictionary<List<bool>, TupleTypeDecl>(new Dafny.ListComparer<bool>());
```

RustanLeino commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,61 @@ public class BuiltIns
+      }
+    }
+
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> argumentGhostness = null) {
```
Is

    Contract.Requires(argumentGhostness == null || argumentGhostness.Count == dims);

the intention?

fpoli replied:
Yes. I added that annotation.

RustanLeino commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,61 @@ public class BuiltIns
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
```
I don't understand this special case. Is it needed?

fpoli replied:
Yes, as a base case for the recursion and to build a key even for empty tuples.

RustanLeino commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4698,35 +4732,47 @@ public enum ES { NotYetComputed, Never, ConsultTypeArguments }
+      get { return ArgumentGhostness.Count; }
+    }
+    
+    public int NonGhostDims
+    {
+      get { return ArgumentGhostness.Count(x => !x); }
+    }
```
This can be written as 
```suggestion
    public int NonGhostDims => ArgumentGhostness.Count(x => !x);
```

in C# (and similarly for the `Dims` getter above).

fpoli replied:
Changed, thanks.

RustanLeino commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -4698,35 +4732,47 @@ public enum ES { NotYetComputed, Never, ConsultTypeArguments }
-      Contract.Requires(0 <= dims);
+    public TupleTypeDecl(List<bool> argumentGhostness, ModuleDefinition systemModule, Attributes attributes)
+      : this(systemModule, CreateCovariantTypeParameters(argumentGhostness.Count), argumentGhostness, attributes) {
+      Contract.Requires(0 <= Dims);
```
Where does this `Dims` come from? It doesn't seem to make sense as a precondition from the caller's perspective. Perhaps you can move this `Contract.Requires` down a couple of lines and instead state it as a `Contract.Assert`?

fpoli replied:
Uh, good catch. The `Dims` method cannot work yet at that point. It should be `argumentGhostness.Count`.

RustanLeino commented on Source/Dafny/Resolver.cs:
```diff
@@ -16928,6 +16926,10 @@ class DefaultValueSubstituter : Substituter {
       return ok && ctor.Formals.Count == dtv.Arguments.Count;
     }
 
+    public static string IsGhostPrefix(bool isGhost) {
```
Would a better name for this method be `GhostPrefix`?

fpoli replied:
I agree. I renamed it.

RustanLeino commented on Test/ghost/AmbiguousGhost.dfy:
```diff
@@ -0,0 +1,16 @@
+// RUN: %diff "%s.expect" "%t"
+
+// Test that the `IsTypeSequence` method of the parser allows tuples with ghost components.
+module M1 {
+  class C1 {
```
The `module` and `class` are allowed, but since they are not used in the example, you might as well remove them.

fpoli replied:
Done.

RustanLeino commented on Test/ghost/AllowedGhost.dfy:
```diff
@@ -0,0 +1,10 @@
+
+method Test()
+{
+    // Tuples with a single ghost component are allowed and they are compiled to empty tuples.
+    var x := (ghost 123);
+    var (y) := (ghost 123);
+    ghost var z: (ghost int) := (ghost 123);
```
Here and in the other Dafny test files below, use 2 spaces for indentation.

fpoli replied:
Done.

fpoli commented on Source/Dafny/Dafny.atg:
```diff
@@ -3821,17 +3839,51 @@ ParensExpression<out Expression e>
+       } else {
+         foreach (var arg in args) {
+           if (arg.FormalParameterName != null) {
+             SemErr(arg.FormalParameterName, "mixing named and positional tuple arguments is not allowed");
```
I changed the parser to use the more permissive logic that you suggested below.

fpoli commented:
Let me know if I should squash together all the commits of this PR.

keyboardDrummer commented:
> Let me know if I should squash all the commits of this PR together.

No need, Dafny's repository only allows merging PRs by squashing them.

keyboardDrummer commented on Source/Dafny/Compiler-Csharp.cs:
```diff
@@ -347,7 +347,8 @@ public CsharpCompiler(ErrorReporter reporter)
       //   ...
       // }
-      var DtT_TypeArgs = TypeParameters(dt.TypeArgs);
+      var nonGhostTypeArgs = SelectNonGhost(dt, dt.TypeArgs);
```
I feel we need a compilation step that removes anything ghost, to prevent having to filter out ghostness in each of the compilers, but I guess that's out of scope for this PR.

RustanLeino replied:
I agree, on both accounts.

keyboardDrummer commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -325,30 +325,52 @@ public class BuiltIns
-        }
-        tupleTypeDecls.Add(dims, tt);
+        // tuple#2 is already defined in DafnyRuntime.cs
+        var attributes = dims == 2 && !argumentGhostness.Contains(true) ? DontCompile() : null;
```
Just checking, shouldn't this condition **be** something like 

`argumentGhostness.Where(x => !x).Sum() == 2` ?

fpoli replied:
Good question. In this PR a Dafny tuple of type `(int, int, ghost int)` is compiled in C# to a `Tuple3OOG<T1, T2>` class, so it's not the same as compiling `(int, int)` and the `!argumentGhostness.Contains(true)` is correct. However, I just noticed that`(int, int, ghost int)` is compiled in Java to a `Tuple2<T1, T2>` class, because that compiler ignores the `TupleTypeDecl::CompileName` implementation (arguably, a bug of the compiler). Moreover, after [this decision](https://github.com/dafny-lang/dafny/pull/1269#issuecomment-877360472) in C# classes like`Tuple3OOG<T1, T2>` and `Tuple3OGO<T1, T2>` have exactly the same implementation and they only differ for the name.

@RustanLeino I remember we decided in a meeting to translate `(int, int, ghost int)` to a class `Tuple3OOG<T1, T2>`, but maybe we should reconsider and translate to `Tuple2<T1, T2>`. If we want the latter, it might be the first case in which different top level declarations gets compiled to the same class in the target language, so I might have to add a check to avoid declaring the same class multiple times.

RustanLeino replied:
Let's leave it as `Tuple3OOG` and `Tuple3OGO` etc. in this PR. I suggest we change these to use `Tuple2` in the same future PR where we make the optimization that `Tuple1G<T>` would turn into just `T`.

keyboardDrummer commented on Source/Dafny/Util.cs:
```diff
@@ -445,4 +445,18 @@ public class DependencyMap
     }
   }
+  
+  class ListComparer<T> : IEqualityComparer<List<T>> {
```
Can you make this work for `IEnumerable<T>` instead of `List<T>` ?

fpoli replied:
Done.

keyboardDrummer commented on Source/Dafny/Verifier/Translator.cs:
```diff
@@ -1384,7 +1384,7 @@ internal class PredefinedDecls {
         Bpl.Variable resType = new Bpl.Formal(ctor.tok, new Bpl.TypedIdent(ctor.tok, Bpl.TypedIdent.NoName, predef.DatatypeType), false);
         Bpl.Function fn;
-        if (dt is TupleTypeDecl ttd && ttd.Dims == 2) {
+        if (dt is TupleTypeDecl ttd && ttd.Dims == 2 && ttd.NonGhostDims == 2) {
```
Should `ttd.Dims == 2` still be in here?

fpoli replied:
Tuples like `(int, int, ghost int)` and `(int, int)` should be translated to different types because we want Boogie to know that instances of the former cannot alias instances of the latter.

fpoli replied:
I'm actually not sure how to observe a difference in the verification time of Dafny. It's a suggestion that I got from Rustan.

RustanLeino replied:
The following assertion should be provable by the verifier, which requires that the verifier treat `(int, int, ghost int)` as a different type than `(int, int)`.

```
method M(a: array<(int, int, ghost int)>, b: array<(int, int)>)
{
  assert a as object != b;
}
```

fpoli replied:
I'm going to add that example, thanks.

Regarding the if condition on `tdd`, we want to only match tuples of size 2 that have no ghost components, because in that case we want to use the `Tuple2` definition provided by `DafnyPrelude.bpl`. Other conditions (`ttd.NonGhostDims == 2` only, `ttd.Dims == 2` only, `ttd.Dims == 2 && ttd.NonGhostDims == 0`) lead to Boogie errors like "Error: use of undeclared function: Tclass._System.Tuple3OOG".

keyboardDrummer commented on Source/Dafny/Verifier/Translator.cs:
```diff
@@ -9416,7 +9416,7 @@ AND  Apply(f,h0,bxs) == Apply(f,h0,bxs)
       if (td is ClassDecl cl && cl.IsObjectTrait) {
         // the type constructor for "object" is in DafnyPrelude.bpl
-      } else if (td is TupleTypeDecl ttd && ttd.Dims == 2) {
+      } else if (td is TupleTypeDecl ttd && ttd.Dims == 2 && ttd.NonGhostDims == 2) {
```
Should `ttd.Dims == 2` still be in here?

RustanLeino replied:
Actually, shouldn't it be `ttd.Dims == 2 && ttd.NonGhostDims == 0`?
Also, the same test appears in one more place in `Translator.cs`, around line 1387.

fpoli replied:
I think `ttd.Dims == 2 && ttd.NonGhostDims == 2` is correct. See [this message](https://github.com/dafny-lang/dafny/pull/1269#discussion_r680798493) in the other conversation.

keyboardDrummer reviewed (commented):
Do we need to update the reference manual as well? Here's a section on tuples: https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-tuple-types. I think we can add an example indicating that the ghost modifier is valid in front of types in tuples.

fpoli commented:
@RustanLeino I fixed the equality support check and added the following test
```
method M<T(==)>(a: T)

method P() {
  var x: (int, int, ghost int);
  M(x); // Disallowed. It's not possible to compile an equality function that also checks ghost components.
}
```
